feat(medusa): Cart and totals improvements

adrien2p opened the pull request:
> Metric: cart completion with 4/5 items took 11 seconds when it takes less than 2 seconds now, it can vary depending on the machines. As part of those 2 sec there is 1 sec for stripe (called 3 times) and few hundreds milliseconds for external tax api call

## Cart/Order performance improvement, more specifically during cart completion

The purpose of that Pr is to drastically reduce the response time of the cart completion strategy.
The improvements made here have also some side effects which improve the performances of other part of the code but 
it was not the main goal of it.

All improvements will be applicable in later PR's to also improve everything related to cart and order end point
and totals calculations.

## Totals calculation

Some improvements have been made to the totals service, in order to not break the existing API right away, it has been
decided by myself (but can evolve) to create a new service `TotalsNewService` where all new methods are implemented and consumed from.

Generally speaking there is no mention of any specific entities such as the Cart or Order entity as the different methods
explicitly describe the expected arguments which makes it flexible and do not enforce to pass a specific entity
but only an object that contains the expected properties. The overall idea is to have a totals service
less coupled to specific entities but just do its job which is to compute some calculations based on the
expected arguments.

The overall changes are the following:

- __getLineItemsTotals__

This method accept a collection of `items` as well as some options

- `includeTax` which only enforce the tax_lines to exist either provided by the item itself or the ones that will be queried internally through the tax provider service if `useExistingTaxLines` is not set to true
- `calculationContext` which is the usual calculation context
- `taxRate` which is only there to support backward compatibility. I choose to keep it but the methods have been separated
- `useExistingTaxLines` which enforce the usage of the tax lines from the items. Can be useful in cases where
the tax_lines already exists or have been already fetched and assigned by the consumer in order to not re fetch them if
it is not necessary

This method is of a higher rank as it will only take care of using the right tax lines and call another method
to compute the totals of each item.

To compute the totals of each item there is now two methods which are described bellow

- __getLineItemTotalsLegacy__

As the name describe, this method calcul the totals for the legacy backward compatibility cases where the `taxRate` is
explicitly provided to be used. It is normally in the case of the order.

- __getLineItemTotals__

Which is used to calculate the totals with the new approach using the tax lines.
In that method the options `includeTax` as been kept only to throw if no tax lines are provided by any of the
possible way, either from the item or explicitly provided

---

-  __getLineItemRefund__

This method take a partial item composed of the necessary properties needed to calculate the refund amount.
It is also possible to provide an optional `taxRate` to calculate it the legacy way where the method will call
the __getLineItemRefundLegacy__ method for that.

- __getLineItemRefundLegacy__

---

Now let's look at the gift card part of the totals below

- __getGiftCardTotals__

This method take now a maximun amount that can be gifted, it is possible to pass either the `giftCards` or the `giftCardTransactions`
as part of the options. If the `giftCardTransactions` are provided then a separate method will be called `getGiftCardTransactionsTotals`
in order to provided the right calculation for that collection otherwise the `giftCards` will be used
to calculate the amount.

- __getGiftCardTransactionsTotals__

This method take care of computing the totals based on the transactions and the region

--- 

- __getShippingMethodsTotals__

This method accept a collection of `shippingMethods` as well as some options

- `includeTax` which only enforce the tax_lines to exist either provided by the method itself or the ones that will be queried internally through the tax provider service if `useExistingTaxLines` is not set to true
- `discounts` 
- `calculationContext` which is the usual calculation context
- `taxRate` which is only there to support backward compatibility. I choose to keep it but the methods have been separated
- `useExistingTaxLines` which enforce the usage of the tax lines from the method. Can be useful in cases where
the tax_lines already exists or have been already fetched and assigned by the consumer in order to not re fetch them if
it is not necessary

This method is of a higher rank as it will only take care of using the right tax lines and call another method
to compute the totals of each method.

To compute the totals of each method there is now two methods which are described bellow

- __getShippingMethodTotalsLegacy__ 

As the name describe, this method calcul the totals for the legacy backward compatibility cases where the `taxRate` is
explicitly provided to be used. It is normally in the case of the order.

- __getShippingMethodTotals__ 

Which is used to calculate the totals with the new approach using the tax lines.
In that method the options `includeTax` as been kept only to throw if no tax lines are provided by any of the
possible way, either from the method or explicitly provided

--- 

## Cart completion

The main purpose of the previous improvements and refactoring was to improve the overall performances
of the cart completion flow. To achieve that, multiple changes needed to be made and will be described bellow.

### Cart service and Order service

Most of the changes in the service are separated in two categories

- Methods signature and refactoring
- Improvements of the decorate totals method

Some of the method was always expecting a cart id and then was querying the expected cart. In some cases
we are already holding the cart and it make sense that we have the flexibility to provide this cart
instead of fetching it again, expect in those cases where we are not sure what to provide and therefore 
we prefer giving the hand to the method to fetch if with the appropriate relations. Therefore for some of the 
service methods it is now possible to either give and id or the entity directly. For the later case it
is much more for the people in need of performance improvement and in that case it is the consumer who must
be sure to pass the entity with the appropriate relations.

Finally, the `decorateTotals` had been reworked in order to provide a better readability, understandability and performances.
To do so, the hidden quadratic and cubic loops have been replaces by the usage of map or object and the overall
loops have been simplified or reduced. We have therefore remove O(n^(2 or 3)) + XO(n) by only XO(n) where X is inferior
to what it was before. Also, we now take advantage of being able to query the tax lines or use the existing one
before calculating the total which gives more flexibility and reduce the number of time we rely on an
external party. For the example, for a cart with 4 items we have now few calls to the external party instead of 64 for the overall flow.
Before those changes, the number a call was constantly increasing with the number of item in the cart, it is not
the case anymore.

The overall decorate totals of the order service have also been improved the same way the cart has been improved and therefore
the readability, understandability and performances have been improved.

## Summary

Some other adjustments have been made along the way in order to improve the usability and simplify the usage of 
all the changes and open the door to more refactoring (but simpler) in the entire code base where we are relying
on the totals, cart and order. Therefore, future PR's will be open to tackle those refactoring and align the code base
to move toward the new ways of calculating the totals and the usage of the new api's. It will increase the overall
performances of everything that is related to the cart or the order. It should be taken into consideration
that the future changes will probably also impact the payment provider if they rely on one of the entities
mention above. Also, the actual totals service could potentially be deprecated until a certain period of time
in favour of the new approached.

FIXES CORE-327

dwene commented:
Woooo, I'm excited to try this out tomorrow! Can you try adding a similar thing with the tax_lines for OrderService::decorateTotals also? I'm seeing some long non-database performance times listing orders as well. Thanks!

dwene commented:
Man, this is awesome, checkout cart went from 10 seconds to 2 seconds!

dwene commented on packages/medusa/src/services/order.ts:
```diff
@@ -1408,6 +1490,136 @@ class OrderService extends TransactionBaseService {
+    const useExistingTaxLines = totalsConfig.useExistingTaxLines
+
+    const [itemsTotals, shippingTotals] = await Promise.all([
+      await this.totalsNewService_.getLineItemsTotals(order.items, {
```
I'm pretty sure if you add the await here, the Promise.all is unnecessary, since it's already awaiting when creating the array.

adrien2p replied:
Ahah dummy me, you are right

dwene commented on packages/medusa/src/strategies/cart-completion.ts:
```diff
@@ -326,6 +156,197 @@ class CartCompletionStrategy extends AbstractCartCompletionStrategy {
+    }
+  }
+
+  protected async handlerTaxLineCreated(
```
typo! `handleTaxLineCreated`

dwene commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,632 @@
+  featureFlagRouter: FlagRouter
+}
+
+export default class totalsNewService extends TransactionBaseService {
```
typically Typescript classes are PascalCased right?

dwene reviewed (approved):
Left a couple of small comments. I tested it using medusa-dev and everything seems to work correctly and fast 🚀!

srindom commented on packages/medusa/src/services/cart.ts:
```diff
@@ -349,7 +235,7 @@ class CartService extends TransactionBaseService {
   }
 
-  private async retrieveNew(
+  async retrieveNew(
```
todo: we should avoid having a `[*]New` method in the public API - can we use the existing `retrieve` as a facade, not sure what we would use for splitting the logic though maybe `select` includes `subtotal | total | etc.` -> use old; otherwise use new. Wdyt?

adrien2p replied:
Yeah pretty agree with you as it introduce a two steps changes to the API which would be good to avoid. I look in your proposal and do something about it 👍

srindom commented on packages/medusa/src/services/cart.ts:
```diff
@@ -1420,35 +1299,35 @@ class CartService extends TransactionBaseService {
+        if (cart.total! <= 0) {
           cart.payment_authorized_at = new Date()
-          return await cartRepository.save(cart)
+          await cartRepository.save({
+            id: cart.id,
+            payment_authorized_at: cart.payment_authorized_at,
+          })
```
todo: we are misssing a `cart.updated` event

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+  featureFlagRouter: FlagRouter
+}
+
+export default class TotalsNewService extends TransactionBaseService {
```
nit: can we rename to `NewTotalsService` - `TotalsNew` a bit odd

srindom commented on packages/medusa/src/services/order.ts:
```diff
@@ -459,27 +536,36 @@ class OrderService extends TransactionBaseService {
+   * @param cartOrId
    */
-  async createFromCart(cartId: string): Promise<Order | never> {
+  async createFromCart(cartOrId: string | Cart): Promise<Order | never> {
```
question: why are we adding support for a full cart object as well?

adrien2p replied:
because sometimes we already have the cart in the hand and if the user knows what he is doing I think he should be able to pass it as we create from cart. I am fine to also go back to passing the cart id only but the idea was to allow to reduce the number of queries if some users (like I did) need to increase their performance. Also, it allows to not surcharge the db when we perform a big number of queries. wdyt?

srindom commented on packages/medusa/src/services/order.ts:
```diff
@@ -1408,6 +1499,185 @@ class OrderService extends TransactionBaseService {
+      )
+    }
+
+    // If we are not forced to use the existing tax lines then fetch the tax lines once and then associate them
+    // to the items and methods copy to avoid fetching them later multiple times and then set useExistingTaxLines to force
+    // the total service to use the tax lines from the items and methods
+    if (!useExistingTaxLines) {
+      const taxLinesMaps = await this.taxProviderService_
+        .withTransaction(manager)
+        .getTaxLinesMap(orderItems, calculationContext)
+
+      orderItems.forEach((item) => {
+        if (item.is_return) {
+          return
+        }
+        item.tax_lines = taxLinesMaps.lineItemsTaxLines[item.id] ?? []
+      })
+      orderShippingMethods.forEach((method) => {
+        method.tax_lines = taxLinesMaps.shippingMethodsTaxLines[method.id] ?? []
+      })
+
+      useExistingTaxLines = true
+    }
```
thought: when the order has been placed we have "locked in" so think we should remove this as we will always be using the existing tax lines - can you think of a case that wouldn't do this?

adrien2p replied:
for example if I create an order that is not placed? The alternative for the orders would be that if one of the items have tax_lines then we use them by default? and we remove the option

srindom commented on packages/medusa/src/services/payment-provider.ts:
```diff
@@ -368,11 +369,14 @@ export default class PaymentProviderService extends TransactionBaseService {
-    cart: Cart & { payment_session: PaymentSession }
-  ): Promise<Payment> {
+  async createPayment(data: {
+    id: string
+    total: number
+    region: Region
+    payment_session: PaymentSession
```
suggestion: wdyt of this instead? 
```suggestion
    cartId: string
    amount: number
    currency_code: string
    payment_session: PaymentSession
```

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+  }
+
+  /**
+   * Calcul and return the items totals for either the legacy calculation or the new calculation
```
```suggestion
   * Calculate and return the items totals for either the legacy calculation or the new calculation
```

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+   * @param lineItemAllocation
+   * @param calculationContext
+   */
+  async getLineItemTotalsLegacy(
```
thought: can't we delete this altogether as this is a "new" service?

adrien2p replied:
I kept it thinking that we might still want to benefit from the improvement even if you use a legacy tax system. Unless we say that we don't support those type of object that use the tax rate anymore. The idea is that the actual totals service will be replaced by the new one once we have refactored everything

srindom replied:
Ah alright - sorry thought this was the old method from the existing TotalsService

adrien2p replied:
ahah no problem, none of the method from this service are the same of the old one, they have been re worked 👍

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+    this.featureFlagRouter_ = featureFlagRouter
+    this.taxCalculationStrategy_ = taxCalculationStrategy
+  }
+
```
suggestion: could we overload `getLineItemTotals` so that we have:
```ts
async getLineItemTotals(items: LineItem[], config: LineItemTotalsConfig): Promise<{ [lineItemId: string]: LineItemTotals }>

async getLineItemTotals(item: LineItem, config: LineItemTotalsConfig): Promise<LineItemTotals>
```

It is confusing that two methods do a similar thing one just takes an array? 

Also wouldn't it make sense to align the two implementations so getLineItemTotals basically is the same as:
```
getLineItemTotals([item], ...)
```

adrien2p replied:
I can just make the `getLineItemTotals` protected and only export the `getLineItemsTotals` as part of the public API. The idea was to separate the calculation from the rest in order to make it  easier to read and have smaller function. wdyt? the `getLineItemTotals` only take care of the calculation

adrien2p replied:
I did make the other method protected and only the main one it public, which accept a collection of items and that can be only one item.

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+   * @param calculationContext
+   * @param useExistingTaxLines Force to use the tax lines of the shipping method instead of fetching them
+   */
+  async getShippingMethodsTotals(
```
thought: Same thought as above with `getLineItemTotals` - can we overload instead?

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+   * @param taxLines
+   * @param discounts
+   */
+  async getShippingMethodTotalsLegacy(
```
can we delete?

adrien2p replied:
probably not, related to https://github.com/medusajs/medusa/pull/2475#discussion_r1012768714

srindom commented on packages/medusa/src/services/totals-new.ts:
```diff
@@ -0,0 +1,692 @@
+   * @param taxRate
+   * @protected
+   */
+  getLineItemRefundLegacy(
```
can we delete?

adrien2p replied:
probably not, related to https://github.com/medusajs/medusa/pull/2475#discussion_r1012768714

srindom commented on packages/medusa/src/services/new-totals.ts:
```diff
@@ -0,0 +1,707 @@
+   * @param taxRate
+   * @param useExistingTaxLines Force to use the tax lines of the line item instead of fetching them
+   */
+  async getLineItemsTotals(
```
todo: rename so the public API is `getLineItemTotals` (change the protected method to something else).

adrien2p replied:
but it takes a collection, it does not make sense to be item totals and not items totals since it is always multiple totals the plural on totals does not represent that. wdyt?

srindom commented on packages/medusa/src/services/new-totals.ts:
```diff
@@ -0,0 +1,707 @@
+   * @param calculationContext
+   * @param useExistingTaxLines Force to use the tax lines of the shipping method instead of fetching them
+   */
+  async getShippingMethodsTotals(
```
todo: like above - rename to `getShippingMethodTotals`

srindom commented on packages/medusa/src/services/order.ts:
```diff
@@ -309,22 +326,63 @@ class OrderService extends TransactionBaseService {
    */
   async retrieve(
-    orderId: string,
+    orderIdOrSelector: string | Selector<Order>,
```
todo: the retrieve method should not allow a selector like this; you can add an internal method if needed otherwise just use `list`?

srindom commented on packages/medusa/src/services/payment-provider.ts:
```diff
@@ -368,11 +368,14 @@ export default class PaymentProviderService extends TransactionBaseService {
-    cart: Cart & { payment_session: PaymentSession }
-  ): Promise<Payment> {
+  async createPayment(data: {
+    cartId: string
```
todo: sorry my previous comment was off - should be snake cased
```suggestion
    cart_id: string
```

srindom commented on packages/medusa/src/services/order.ts:
```diff
@@ -459,27 +551,36 @@ class OrderService extends TransactionBaseService {
-          "items.adjustments",
-        ],
-      })
+      const exists = !!(await this.retrieve({
+        cart_id: isString(cartOrId) ? cartOrId : cartOrId?.id,
+      }).catch(() => false))
```
```suggestion
      const exists = !!(await this.retrieveByCartId(isString(cartOrId) ? cartOrId : cartOrId?.id, { select: ["id"] })
        .catch(() => false))
```

srindom reviewed (commented):
A few more fixes and we should be ready to go 👍

srindom commented on packages/medusa/src/strategies/cart-completion.ts:
```diff
@@ -299,19 +301,12 @@ class CartCompletionStrategy extends AbstractCartCompletionStrategy {
-              relations: ["shipping_address", "payments"],
-            }
-          )
+        order = await orderServiceTx.retrieveByCartId(id, {
+          relations: ["shipping_address", "payments"],
+        })
```
thought: do we need to decorate totals here too?

adrien2p replied:
No, I need to find a way to split that method to get the raw result but i am on another thing at the moment 👍

srindom reviewed (commented):
Looks good - a few tests and we can merge :)

adrien2p commented:
@srindom @olivermrbl 
- I ve added the new unit tests suite for the new totals service (this is why the PR look enormous now, a lot of tests 😹). 
- I have removed the notion of `useExistingTaxLines` in order to simplify the api usage. Since the getTaxLines is now properly manage it does not add a lot of overhead so it is a trade off.
- I have re applied the changes made to the get-order end point and cleaned up the legacy decorate totals
- various clean up
- I did add some todo for the places where we are using entire entities in the creation process which can lead to errors with cascade insert if some data should not be inserted but are present (tax lines on the items for example)

I suggest to re review the PR just to be sure :)

patrick-medusajs commented on packages/medusa/src/services/order.ts:
```diff
@@ -570,18 +638,55 @@ class OrderService extends TransactionBaseService {
+        )
+      }
+
+      // gift cardable amount
+      /* const calculationContext = await this.totalsService_.getCalculationContext(
+         cart,
+         {
+           exclude_gift_cards: true
+         }
+       )
+       const totals = await this.newTotalsService_.getLineItemTotals(cart.items, {
+         includeTax: true,
+         calculationContext
+       })
+ 
+       let gcBalance = 0
+       if (cart.region?.gift_cards_taxable) {
+         const [subtotal, discountTotal] = Object.values(totals).reduce((acc, total) => {
+           acc[0] += total.total
+           acc[1] += total.discount_total
+           return acc
+         }, [0, 0])
+ 
+         gcBalance = subtotal - discountTotal
+       } else {
+         gcBalance = Object.values(totals).reduce((acc, total) => {
+           acc += total.total
+           return acc
+         }, 0)
+       }*/
```
to remove?

adrien2p replied:
Nice catch 😅 it was a test of my own ahah

patrick-medusajs commented on packages/medusa/src/services/new-totals.ts:
```diff
@@ -0,0 +1,738 @@
+   * @param includeTax
+   * @param calculationContext
+   * @param taxRate
+   * @param useExistingTaxLines Force to use the tax lines of the line item instead of fetching them
```
useExistingTaxLines is not part of the method signature

adrien2p replied:
Nice catch forgot to update the jsdoc

patrick-medusajs commented on packages/medusa/src/services/new-totals.ts:
```diff
@@ -0,0 +1,738 @@
+   * @param shippingMethod
+   * @param calculationContext
+   * @param taxLines
+   * @param discounts
```
does not match method signature

adrien2p replied:
It is auto generated but it seams that tax rate is missing nice catch thanks 🙌

patrick-medusajs reviewed (commented):
Did my best to contribute but this PR exceeds my level of familiarity with the code base.
That being said, it was a great opportunity to get expose to this area of the system.

adrien2p commented on packages/medusa/src/services/__mocks__/new-totals.js:
```diff
@@ -0,0 +1,42 @@
+  },
+  getLineItemsTotals: jest.fn().mockImplementation(() => {
+    return Promise.resolve(new Map())
+  }),
```
It was a request from seb to be singular instead of plural. Tried to arguments 
 but couldnmt arguments more

patrick-medusajs replied:
Let me rephrase, both getLineItemsTotals and getLineItemTotals are declared on the mock but only getLineItemTotals is declared on the actual NewTotalsService.

adrien2p replied:
Yes, cause only that method exists now. I also have simplified the totals api 👍

patrick-medusajs replied:
Why mock a method that doesn't exits? https://github.com/medusajs/medusa/blob/182d2378fcd87bfcd1273240989ca2315b31c1e1/packages/medusa/src/services/new-totals.ts

adrien2p replied:
You are right, it use to exists but i should remove it noce catch 🙏 i ll clean it up

adrien2p commented on packages/medusa/src/services/__mocks__/new-totals.js:
```diff
@@ -0,0 +1,42 @@
+    .mockImplementation((order, lineItems) => {
+      return Promise.resolve({})
+    }),
+  getCalculationContext: jest.fn().mockImplementation((order, lineItems) => {
+    return Promise.resolve({})
+  }),
```
Cause the migration of that method is not part of the scope of that pr 👌keep it for a later pr

adrien2p commented on packages/medusa/src/services/__mocks__/new-totals.js:
```diff
@@ -0,0 +1,42 @@
+  getLineItemTotals: jest.fn().mockImplementation(() => {
+    return Promise.resolve({})
+  }),
+  getOrderLineItemTotals: jest.fn().mockImplementation(() => {
+    return Promise.resolve({})
+  }),
```
Same as the first comment

adrien2p replied:
There is no more references to specific entities such as cart or order since we are using items in reality . It is not a migration of the existing but a re work.

adrien2p commented on packages/medusa/src/services/__mocks__/new-totals.js:
```diff
@@ -0,0 +1,42 @@
+  getCalculationContext: jest.fn().mockImplementation((order, lineItems) => {
+    return Promise.resolve({})
+  }),
+  getShippingMethodsTotals: jest.fn().mockImplementation((order, lineItems) => {
+    return Promise.resolve(new Map())
+  }),
```
Same as previous 👌

adrien2p commented on packages/medusa/src/services/__mocks__/new-totals.js:
```diff
@@ -0,0 +1,42 @@
+  getShippingMethodTotals: jest.fn().mockImplementation((order, lineItems) => {
+    return Promise.resolve({})
+  }),
+  getOrderShippingMethodTotals: jest
+    .fn()
+    .mockImplementation((order, lineItems) => {
+      return Promise.resolve({})
+    }),
+}
```
Same as previous about the re work

olivermrbl commented on packages/medusa/src/services/cart.ts:
```diff
@@ -315,66 +200,74 @@ class CartService extends TransactionBaseService {
 
-    if (relations && relations.length > 0) {
-      query.relations = relations
+    if ((options.select || []).length <= 0) {
```
suggestion(non-blocking): 
```suggestion
    if ((options.select || []).length === 0) {
```

olivermrbl commented on packages/medusa/src/services/cart.ts:
```diff
@@ -1413,35 +1306,34 @@ class CartService extends TransactionBaseService {
-    context: Record<string, unknown> = {}
+    context: Record<string, unknown> & {
+      cart_id: string
+      useExistingTaxLine?: boolean
```
suggestion(non-blocking): `snake_case`

olivermrbl replied:
question: Is this context option used? If not, let's get rid of it

adrien2p replied:
I removed the useExisting... but the context is passed to authorizePayment

olivermrbl commented on packages/medusa/src/strategies/cart-completion.ts:
```diff
@@ -326,6 +156,182 @@ class CartCompletionStrategy extends AbstractCartCompletionStrategy {
+        ...context,
+        cart_id: id,
+        idempotency_key: idempotencyKey,
+        useExistingTaxLine: true,
```
question: I don't think we are using this option any longer. Can you confirm or deny?

adrien2p replied:
we don't use it anymore I ve simplified

olivermrbl commented on packages/medusa/src/strategies/cart-completion.ts:
```diff
@@ -326,6 +156,182 @@ class CartCompletionStrategy extends AbstractCartCompletionStrategy {
+      }
+    }
+
+    if (!isDefined(cart.total)) {
+      return {
+        response_code: 500,
+        response_body: {
+          message: "Unexpected state",
+        },
+      }
+    }
```
thought: I think this seems a bit arbitrary. Can we add a comment and improve the error message to provide a better experience for both developers reading the code as well as the ones who are consuming the API?

adrien2p replied:
I should be able to remove it as it is a legacy error which should not happen anymore with the new decorate totals method

olivermrbl commented on packages/medusa/src/strategies/cart-completion.ts:
```diff
@@ -4,14 +4,17 @@ import { EntityManager } from "typeorm"
 import { IdempotencyKey, Order } from "../models"
```
praise: Really like the way you've refactored the strategy into isolated methods

olivermrbl commented on packages/medusa/src/services/swap.ts:
```diff
@@ -802,7 +796,13 @@ class SwapService extends TransactionBaseService {
       swap.shipping_address_id = cart.shipping_address_id
-      swap.shipping_methods = cart.shipping_methods
+      // TODO: Due to cascade insert we have to remove the tax_lines that have been added by the cart decorate totals.
+      // Is the cascade insert really used? Also, is it really necessary to pass the entire entities when creating or updating?
```
comment: I think it would be appropriate to remove the cascade on shipping method in the swap model, but can be left for a follow-up iteration 👍

adrien2p replied:
Yeah it is what I thought, I keep the todo for later refactoring to not make the pr bigger

olivermrbl commented on packages/medusa/src/services/order.ts:
```diff
@@ -43,6 +44,9 @@ import RegionService from "./region"
 import TotalsService from "./totals"
+import { NewTotalsService, TaxProviderService } from "./index"
+
+export const ORDER_CART_ALREADY_EXISTS_ERROR = "Order from cart already exists"
```
thought(non-blocking): Should we change this to an error type or code to not rely on a correct error message in the strategy? 

If so, I think we should move it to `medusa-core-utils/src/errors.ts`

olivermrbl commented on packages/medusa/src/services/cart.ts:
```diff
@@ -315,66 +200,74 @@ class CartService extends TransactionBaseService {
   }
 
-  private async retrieveNew(
+  protected async retrieveLegacy(
```
suggestion(non-blocking): We could add `@deprecated` if that's the case

adrien2p commented on packages/medusa/src/services/cart.ts:
```diff
@@ -1428,27 +1430,18 @@ class CartService extends TransactionBaseService {
+            id: cart.id,
+            payment_authorized_at: cart.payment_authorized_at,
+          })
+          return cart
```
I don't understand, the cart has been saved and we hold the cart. there is an action in the middle (emit event) I think you are not up to date.

olivermrbl replied:
Sorry, this was an old comment from last week who got included in the review. Was def. not up to date :)

olivermrbl commented:
@srindom Do you have any more comments on this one? Let us get it merged, deployed to staging, and tested asap 💪

srindom reviewed (approved):
LGTM!
